### PR TITLE
[None][fix] Refactor triton paddings

### PR DESCRIPTION
**Pull request by dongfengy**
## Summary by CodeRabbit

- New Features
  - Unified padding/sharding workflow for tensor-parallel fused MoE and added bias support in weight creation.
  - Expanded MXFP4/FP8 weight and scale layout support for broader hardware/configuration compatibility.

- Refactor
  - Replaced ad-hoc padding with configurable, consistent padding multiples and a shard-and-pad preparation flow.
  - Weight/scale preparation standardized across loading and activation paths.

- Bug Fixes
  - Stricter activation/output padding and scale alignment to prevent dimension mismatches and improve reliability.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 1]
```diff
@@ -671,48 +723,55 @@ def __init__(self, activation_dtype):
         assert activation_dtype in [torch.float8_e4m3fn, torch.bfloat16], \
             f"TritonMXFP4FusedMoEMethod only supports float8_e4m3fn or bfloat16 activation, got {activation_dtype}"
         self.activation_dtype = activation_dtype
-        self.in_dim_padding_multiple = 128
-        self.out_dim_padding_multiple = 256
+
+        self.k_padding = 128
+        self.n_padding = 2 * self.k_padding
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py]: Done

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 2]
```diff
@@ -663,6 +660,61 @@ def swizzle_weight_and_scale(w: torch.Tensor, w_scale: torch.Tensor):
...
+                                      padding[non_shard_axis])
+        size_to_pad[non_shard_axis] = padded_size - tensor.shape[non_shard_axis]
+
+    # Actually call pad
+    if any(size_to_pad):
+        pad = (0, size_to_pad[0]) if tensor.dim() == 1 else (0, size_to_pad[1],
+                                                             0, size_to_pad[0])
+        tensor = torch.nn.functional.pad(tensor, pad)
+
+    return tensor.to(device)
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py]: Done

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 3]
```diff
@@ -830,20 +889,29 @@ def load_expert_w3_w1_weight(self,
...
-                                            device=device)
-        w3_weight_shard = load_weight_shard(w3_weight,
-                                            module.tp_size,
-                                            module.tp_rank,
-                                            TensorParallelMode.COLUMN,
-                                            device=device)
+        # Use full k-padding for float tensors, half for already-packed uint8
+        k_pad = self.k_padding if w1_weight.dtype in (
+            torch.bfloat16, torch.float16,
+            torch.float32) else self.k_padding // 2
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py]: I have switched the if else order as you suggested. But I don't think we need to involve quant_config here. The check is mainly because of we share the same function for weights and biases but biases are not mxfp4.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 4]
```diff
@@ -889,12 +957,16 @@ def load_expert_w2_weight(self,
...
         assert device.type == "cuda"
-        w2_weight_shard = load_weight_shard(w2_weight,
-                                            module.tp_size,
-                                            module.tp_rank,
-                                            TensorParallelMode.ROW,
-                                            device=device)
-
+        k_pad = self.k_padding if w2_weight.dtype in (
+            torch.bfloat16, torch.float16,
+            torch.float32) else self.k_padding // 2
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py]: answered above.

**Review by yuxianq — APPROVED** [review]
